### DISCUSSION TITLE
How do we compute averaged quantities for a pixel footprint in a BSDF? [mipmapping][ray-differentials]

### DISCUSSION THREAD

**Discussion by Vrroom** [mitsuba-renderer/mitsuba3#1175] [Q&A]
Hello All, 

I was wondering how do we calculate, say, an average normal, for a pixel footprint in a BSDF. Currently, the surface interaction construct gives us access to dp_du and dp_dv where, based on my understanding, p is a point in world space. 

However, in order to compute average normal in a pixel footprint, we need access to du_dx and du_dy where x and y are the pixels. Is there a simple change-of-variables to transform the available quantities to the ones we need? 

Thanks a lot and looking forward to hearing from you!!

With Regards, 
Sumit.

**Comment by ziyi-zhang** [comment]
Hi,

To compute the "average normal in a pixel footprint", is this equivalent to the normal AOV rendering (https://mitsuba.readthedocs.io/en/latest/src/generated/plugins_integrators.html#arbitrary-output-variables-integrator-aov) with box rfilter?
If not, then we can still do it in a custom integrator.

Best

**Comment by Vrroom** [comment]
Thanks Ziyi, I need to compute the "average normal in a pixel footprint" in my BSDF `sample` function which is why I don't think I need AOV rendering since that'll just give me an image from the point of view of the camera and I'm not sure how to use that to actually use in the `sample` function.

I hope my concern makes sense. Thanks a lot!!

**Comment by Vrroom** [comment]
Hi Ziyi, kindly following up on my above comment. Do you have ideas for how to do the above? I need it to compute a meso BSDF that takes account of visibility (in the presence of micro geometry) and would like to compute it in my `sample` function.

**Reply by njroussel** [reply]
Hi @Vrroom 

Just to clarify, you want the average normal in a camera pixel, or a texture texel on some shape's material?

For the former, what Ziyi recommends makes sense for the first bounce, you'd still need to read  (`dr.gather()`) from the AOV image in your `sample` method. It requires some fairly straightforward indexing, you should look at how the rays are generated for the details. For later bounces, this would not work.
Theoretically, I think you could also use the partials to analytically get the pixel-local normal (the change of variables that you suggest). However, in practice, I don't think we properly handle `duv_dx`/`duv_dy` in most BSDFs. so I wouldn't recommend using that.

**Reply by Vrroom** [reply]
Hi @njroussel, thanks for the response. 

I'll just briefly elaborate my problem setting. My scene description consists of the following:

* A macro geometry (such as a cylinder)
* A micro geometry (normal map, height field and tangent map) that is used to parameterize a Spatially Varying BSDF.

Note that the micro geometry will mask certain regions of the object. To reason about this, we precompute a visibility function for each point on our cylinder.

Now, our goal is to compute an effective BSDF by combining a standard micro BSDF (let's take the Cook Torrance one as an example) with the visibility function. To do this calculation, for each pixel, I need to calculate the average **visible** normal over the pixel footprint. 

If I were to write an AOV to precompute this as you suggest, I don't know how to filter out the invisible normals. Based on my understanding, I can set the active flag to `False` and weight to `(0, 0, 0)` in my `sample` method but that would just mix __black__ color with my normals in the final output image?

Hope my question makes sense. Thanks for taking out the time to reply -- I know you guys are working on a lot of cutting edge stuff!
